### DISCUSSION TITLE
`usePathname` not providing the actual pathname present in the URL

### DISCUSSION THREAD

**Discussion by alessandrojcm** [vercel/next.js#52315] [Help]
### Summary

Hi there!

I am having an issue with the `usePathname` hook, but I am not sure if it is a bug with the hook itself or if it is just some sort of race condition I am not able to debug.

My use case is as follows: we are using [Search UI](https://docs.elastic.co/search-ui) to implement an Amazon-like search bar at work. When the user starts typing in the search bar they'd get autotomplete results but upon hitting enter they would be redirected to a route showing the results (`/search-results` in my example), I went through Search UI [NextJS integration guide](https://docs.elastic.co/search-ui/guides/nextjs-integration) and adapted it to the new router (they use the old `next/router`, see `app/Shell.client.tsx` in my repro; basically it maintains the state writing search params to the URL).

My issue is that, according to [Nextjs docs](https://nextjs.org/docs/app/api-reference/functions/use-pathname#do-something-in-response-to-a-route-change) using `usePathname` and `useSearchParams` wrapped inside an `useEffect`/`useMemo`/`useCallback` should provide you with the up to date pathname but this is not happening. What happens is that, once the user hits enter and gets navigated to `/search-results`, Search UI will try to update the state in the URL but `usePathname` will still be returning the index path; so the user will be brieflly navigating to `/search-results` and then inmediatly will be navigated back to `/`, but if I stop with the debugger and check the `window.location` object then it gives me back the actual URL.

Is there something I am missing here?

UPDATE: kinda worked around it with https://github.com/vercel/next.js/discussions/42016#discussioncomment-5452267 but I am not sure if this is the best solution

### Additional information

```js
Operating System:
      Platform: win32
      Arch: x64
      Version: Windows 10 Enterprise
    Binaries:
      Node: 19.7.0
      npm: N/A
      Yarn: N/A
      pnpm: N/A
    Relevant Packages:
      next: 13.4.8
      eslint-config-next: 13.4.8
      react: 18.2.0
      react-dom: 18.2.0
      typescript: 5.1.6
    Next.js Config:
      output: N/A
```

### Example

https://github.com/alessandrojcm/usepathname-repro

**Comment by GuiLous** [comment]
Im facing the same problem. In my case i have a page called "/jogar" and another page "/manutencao". When the user click to login and the back-end is under maintenance the user is redirect to "/manutencao" with router.push("/manutencao"). The problem is: On my "maintenance" page i have a button to redirect back to Home "/" but when the user click on the button i check the "pathname" from "usePathname" but the "pathname" is getting me "/jogar" instead of "/manutencao" so the redirect goes to the wrong page.

**Reply by alessandrojcm** [reply]
Supposedly this is to be avoided via passing the path names as the dependency array of a hook (be it memo, callback or effect) but that hasn't work for me.
